### ISSUE TITLE
Need a consistent way to manage switching out a custom operator catalog as clusters are being upgraded to newer versions

### ISSUE THREAD

**Issue by huizengaJoe** [operator-framework/operator-lifecycle-manager#1769]
## Feature Request

**Is your feature request related to a problem? Please describe.**

Problem Statement:

- As a **Consumer** of Operators delivered in a custom catalog, I do not want to manually update my catalog to point to a new index image for each OCP release.  I want the catalog to automatically update and work with my version of OCP installed.
- As a **Consumer** of Operators delivered in a custom catalog, I would like the ability to verify "prior to enabling" a new OCP specific catalog, that the new catalog will not break or remove operators that my running workloads depend on. I would like some sort of "dry-run" or "look ahead" capability to help me with an OCP upgrade that affects my catalog and its operators.
- As a **Developer** of Operators delivered in a custom catalog, I want to support multiple versions of OCP that can install my software through OLM. 
- As a **Developer** of Operators delivered in a custom catalog, I want to use features of olm/opm/ocp in my catalog index that may not be available in previous versions of OCP.

**Describe the solution you'd like**

Not sure I know what the correct solution is, but would be great if we collaborated on a solution that would work for any product that is delivered using a custom operator catalog.

It would be interesting to understand how OCP manages this catalog switching process for the 4 internal catalogs it creates and manages, maybe there is something that could be externalized to help with custom catalog management.

We had also kicked around the idea of a catalog manager operator. It could watch for changes to an ocp version and then take action, maybe produce some sort of "look ahead/dry run" summary...

**Comment by exdx** [comment]
What are your concerns for a catalog when OCP is upgraded? The OCP pipeline extensively validates each operator version in the new version of OCP, so those default catalogs should work. For custom catalogs, doing that same kind of validation could be pretty labor intensive, I can see that. Is there a specific case where upgrading OCP broke a custom catalog you were using? 

Regarding updating the index image, check out the new catalog polling feature: https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/catalog-polling.md. This is how OCP-provided catalogs (like red hat operators) are getting updates in newer versions of OCP.

**Comment by ecordell** [comment]
> We had also kicked around the idea of a catalog manager operator. It could watch for changes to an ocp version and then take action, maybe produce some sort of "look ahead/dry run" summary...

This is probably the simplest / most immediate solution to the problem. It's how OpenShift handles this as well (the cluster-version-operator sets the image pullspecs for the default CatalogSources, albeit indirectly through the marketplace-operator).

Other ideas that are more general / attractive, but which don't exist today and will take some time and thought:

- Add a non-image-polling option to find updates for catalogs that can hit an endpoint with metadata about the current cluster and receive a response with the correct target image spec. OpenShift does something similar by hitting a [cincinatti](https://github.com/openshift/cincinnati) endpoint.
- Run a proxy in front of the registry that contains the catalog, make decisions about what actual image content to serve based on the request (auth token, headers, etc).
- Templating the imagepull spec based on the cluster environment

**Comment by huizengaJoe** [comment]
@ecordell I have starting thinking a lot more about this in the last week and also starting working on a simple golang prototype, 
here are the types of things I am looking at, I started down this path as a cli and am still not sure how it should manifest itself in a cluster.  Should it be an operator, should it be some sort of extension/plugin to OLM.....

```
Available Commands:
  checkCluster          Command used to collect runtime details about a cluster
  compareCatalogs       Command used to compare 2 catalogs
  createCatalog         Command used to install an operator catalog
  deleteCatalog         Command used to remove an operator catalog
  dumpCatalog           Command used to dump an operator catalog
  getBundleThatReplaces Command used to check if new catalog will drive an update to existing subscription
  getCorrectCatalog     Command used to Idenitify the correct catalog that should be running in a cluster
  getCurrentCSV         Command used to get the current CSV for a given operator
  help                  Help about any command
  listSubscriptions     Command used to list all of the subscriptions currently using a catalog
  manageCatalog         Command used to instantiate correct catalog based on runtime attributes of a cluster
  ```

**Comment by huizengaJoe** [comment]
@dmesser you mentioned in call today that there is another issue related to this one where is sounded like there were additional values we could add to a catalogSource that could be used to switch which catalog is running, can you reference that issue here please?

We also talked about maybe adding another updateStrategy. It would be interesting to have some sort of generic pluggable way to deliver an updateStrategy implementation.

The whole catalog management set of use cases is also something that would be great to flush out

**Comment by dmesser** [comment]
@dmesser I don't think we have an upstream issues for this yet, but the idea was that catalog references could contain variables, that align with platform versioning or other aspects of the platform (architecture, brand etc). So as a specific example, a user would be able to say that the catalog image would come from: `quay.io/acme/my-catalog:v{ocp-major}.{ocp-minor}` or  `quay.io/acme/my-catalog:v{k8s-version}`. OLM would reconcile those and as a result a catalog change could be triggered when the platform is updated.

**Comment by jchunkins** [comment]
Referencing the conversation that took place in slack about this topic: https://kubernetes.slack.com/archives/C0181L6JYQ2/p1622730604039200

Explicit example provided by @ecordell (which is similar to what @dmesser suggested in previous comment):


if you could write a catalogsource with:
```
image: quay.io/my/catalog:{ocp_major_minor}
```
or
```
image: quay.io/my/catalog:{kube_major_minor}
```

to get the catalog pod to reference `quay.io/my/catalog:4.7` or `quay.io/my/catalog:1.19`

**Comment by dmesser** [comment]
@jchunkins This issue is tracked here in OLMs backlog: https://issues.redhat.com/browse/OLM-2011

As you can see it's medium priority for the team due to other priorities (declarative config, new Bundle API) etc. I think the proposed implementation directly in the catalog operator would make for a nice, reasonably isolated and well-scoped feature that upstream contributors could take up. WDYT?

**Comment by huizengaJoe** [comment]
@dmesser @jchunkins I am still a bit fuzzy on the overall process to attempt to proceed with this work. Where should we be working to flush out a design for this templating work? Here in this repo? and when/what should go in the jira issue?

One of the first questions I have,  what is the best way to determine the ocp major, minor versions?  In my own sandbox I was using the clusterversion resource status.history and ordering by semver but guessing there may be a better way

**Comment by dmesser** [comment]
@huizengaJoe At a high level you would follow the contribution guide: https://github.com/operator-framework/operator-lifecycle-manager/blob/master/CONTRIBUTING.md 

In this particular instance, the initial requirements/ideas are defined in the [Jira](https://issues.redhat.com/browse/OLM-2011). Usually what the OLM team does is create an enhancement design following this process here https://github.com/operator-framework/enhancements/ which should be kept to a minimum to lay out the desired end state and drive a prototype. The latter you could share in a demo in the OLM WG meetings based on your fork. That's where you'd also discuss some implementation details like determining OCP major/minor version.

**Comment by timflannagan** [comment]
Should we close this issue out now that John's catalog templating work merged in the v0.19.x minor version release? It looks like the Epic Daniel linked it also in the closed status, so I just wanted to double check whether there's an existing work we wanted to tackle, or whether the templating logic was sufficient.

**Comment by jchunkins** [comment]
@timflannagan Thanks for pointing this out... I agree that this issue has been implemented and we can close this.

/close
